### PR TITLE
Fix EDF+ annotation timestamps to include millisecond aspect of recording starttime

### PR DESCRIPTION
**Pull request by greydongilmore**

### CODE REVIEW

File: fileio/ft_read_event.m [hunk 1]
```diff
@@ -678,7 +678,25 @@
...
+      % header. If X=0, then the .X may be omitted. Onset must start
+      % with a '+' or a '-' character and specifies the amount of
+      % seconds by which the onset of the annotated event follows
+      % ('+') or precedes ('-') the startdate/time of the file,
+      % that is specified in the header.
+
+      % determine millisecond aspect of starttime (always within
+      % first data record), remove from remaining timestamps
+      tok = tokenize(tal{1}, char(20));
+      millisecond_start = -1*str2double(tok{1});
```
Comment (schoffelen) [inline fileio/ft_read_event.m:698]: from the description it seems that there are cases where there should be an addition '+' or a subtraction '-', it is not clear to me whether the -1 is something that needs to be always applied, or whether it depends on an addition '+'/'-'.

File: fileio/ft_read_event.m [hunk 2]
```diff
@@ -687,11 +705,11 @@
         if any(tok{1}==21)
           % the time and duration are specified
           dum = tokenize(tok{1}, char(21)); % split time and duration
-          time     = str2double(dum{1});
-          duration = str2double(dum{2});
+          time     = str2double(dum{1}) + millisecond_start;
+          duration = str2double(dum{2}) + millisecond_start;
```
Comment (schoffelen) [inline fileio/ft_read_event.m:709]: why does duration need adjustment?

File: fileio/ft_read_event.m [hunk 3]
```diff
@@ -687,11 +705,11 @@
         if any(tok{1}==21)
           % the time and duration are specified
           dum = tokenize(tok{1}, char(21)); % split time and duration
-          time     = str2double(dum{1});
-          duration = str2double(dum{2});
+          time     = str2double(dum{1}) + millisecond_start;
```
Comment (schoffelen) [inline fileio/ft_read_event.m:708]: are you sure that this needs to be a '+'?

In the example that you provide on the github issue, it seems that a positive number needs to be added here, in order to get the numbers that FieldTrip returns sync'ed to the ones in the EDF editor. However, I don't know what happens when you do the -1 multiplication a few lines up. Please check or confirm that this is robust enough, and works always.

File: fileio/ft_read_event.m (hunk 2 above)
Comment (greydongilmore) [reply fileio/ft_read_event.m:709]: In almost all instances duration will not be specified, meaning only an onset was defined for the annotation. However, the duration could take on a real value indicating how long the annotation signal went till. The onset/duration relationship should be considered as (onset, till). So onset indicates when the annotation signal began in seconds from the start (minus the millisecond aspect). Duration would indicate at which time the annotation ended from the start of the recording (minus the millisecond aspect). If only removed the millisecond value from the onset you widen the gap between onset and duration (which are both timestamps relative to the start).

File: fileio/ft_read_event.m (hunk 3 above)
Comment (greydongilmore) [reply fileio/ft_read_event.m:708]: I think the first comment covers this.

File: fileio/ft_read_event.m (hunk 2 above)
Comment (schoffelen) [reply fileio/ft_read_event.m:709]: OK, thanks for the clarification. But in Fieldtrip the 'duration' actually means the 'length' of the event, i.e. the temporal difference between its onset (as indicated by time), and its 'offset', which is not explicitly represented. Does this mean then that the actual correct specification of duration should be str2double(dum{2})-str2double(dum{1}); ?

File: fileio/ft_read_event.m (hunk 1 above)
Comment (schoffelen) [reply fileio/ft_read_event.m:698]: I see. I think I misunderstood the whole '-'/'+' discussion. This refers to the onset times of the subsequent annotations; so the first 'millisecond' thingy is always positive valued.

File: fileio/ft_read_event.m (hunk 2 above)
Comment (schoffelen) [reply fileio/ft_read_event.m:709]: hmmm, according to the documentation, it seems that duration indeed is the length of the event, see paragraph 2.2.2, so I guess the millisecond_start should go

**Comment by schoffelen** [comment]
I will merge this into my branch for now, so that I can run some tests locally. Currently, our fileserver is down for maintenance, so I can only test a bit more extensively later today, or if we are unlucky somehwere tomorrow.

**Comment by greydongilmore** [comment]
@schoffelen onset and duration are both relative to the `starttime` of the file. This would be an example annotation from a data block in the TAL channel:

> onset: 420sec 
duration: 450sec
event: eyes closed

If you wanted to get the length of the event you would subtract `duration` from `onset`:

> length: 450-420 = 30.000 sec 

If the millisecond offset is not removed from `duration` then the situation would be:

>onset: 420sec +(-1*.100sec)
duration: 450sec
event: eyes closed
length: 30.100 secs

You can see that the gap between the two events would widen, adding time to the length of the event.

**Comment by schoffelen** [comment]
Thanks Greydon, and sorry for my thickheadedness. I think I now merged this branch as is, so all should be fine now. Thanks for your contribution and patience.

**Comment by greydongilmore** [comment]
no worries at all, I am prepping for my Ph.D. comp exam for this Tuesday so I apologize for not being as responsive to this PR. Glad we got it figured out.

**Comment by schoffelen** [comment]
Go for it! Good luck.

**Pull request merged**